### PR TITLE
SDK: Integrate PriceObserver

### PR DESCRIPTION
**Pull request by gasiortomasz**

### CODE REVIEW

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 1]
```diff
@@ -1,33 +1,82 @@
 import {BalanceObserver} from './BalanceObserver';
 import {TokenDetailsWithBalance} from '@universal-login/commons';
 import {utils} from 'ethers';
+import {} from '@universal-login/commons/lib';
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: ?

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 2]
```diff
@@ -1,33 +1,82 @@
...
+  private unsubscribePriceObserver?: Function;
+  private unsubscribeBalanceObserver?: Function;
+  private callbacks: Function[] = [];
   constructor(private balanceObserver: BalanceObserver, private priceObserver: PriceObserver) {}
 
-  subscribe(callback: Function, currencySymbol: string) {
-    const callbackWrapper = async (tokensDetailsWithBalance: TokenDetailsWithBalance[]) => {
-      const aggregatedBalance = await this.getAggregatedBalance(tokensDetailsWithBalance, currencySymbol);
-      callback(aggregatedBalance);
+  subscribe(callback: Function) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: Introduce type

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 3]
```diff
@@ -1,33 +1,82 @@
...
+import {PriceObserver} from './PriceObserver';
+import clonedeep from 'lodash.clonedeep';
 
-export interface PriceObserver {
-  getTokenPrice: (tokenSymbol: string, currencySymbol: string) => Promise<number>;
-}
+export type SymbolToValue = {[symbol: string]: number};
 
 export class AggregateBalanceObserver {
+  private tokenPrices: {[symbol: string]: SymbolToValue} = {};
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: Add type `tokenPrices`

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 4]
```diff
@@ -1,33 +1,82 @@
...
+        this.unsubscribePriceObserver!();
+        this.tokenDetailsWithBalance = [];
+        this.tokenPrices = {};
+      }
     };
+    return unsubscribe;
+  }
 
-    return this.balanceObserver.subscribe(callbackWrapper.bind(this));
+  priceObserverCallback(tokenPrices: {}) {
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: ... 
use type `tokenPrices`

File: universal-login-sdk/test/fixtures/PriceObserver.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import {PriceObserver} from '../../lib/core/observers/PriceObserver';
+import {SdkConfigDefault} from '../../lib/config/SdkConfigDefault';
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/fixtures/PriceObserver.ts]: `/fixtures/` -> `/mock/`

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 6]
```diff
@@ -1,33 +1,82 @@
...
-      async (total, tokenDetails) => (await total + await this.getTokenBalance(tokenDetails, currencySymbol)),
-      Promise.resolve(0)
-    );
+  balanceObserverCallback(tokenDetailsWithBalance: TokenDetailsWithBalance[]) {
+    this.tokenDetailsWithBalance = tokenDetailsWithBalance;
+    this.notifyListeners();
   }
 
-  async getTokenBalance(tokenDetailsWithBalance: TokenDetailsWithBalance, currencySymbol: string) {
+  notifyListeners() {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: refreshPrices() or recompute()

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 7]
```diff
@@ -1,33 +1,82 @@
...
+    if (Object.keys(this.tokenPrices).length === 0 || this.tokenDetailsWithBalance.length === 0) {
+      return;
+    }
+    const totalWorth = this.getAggregatedTotalWorth();
+    this.callbacks.forEach((callback) => callback(totalWorth));
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: extract `notifyListeners()`

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 8]
```diff
@@ -1,33 +1,82 @@
...
+    const tokenPricesCopy = clonedeep(tokenPrices);
+    Object.keys(tokenPricesCopy).map((symbol) => tokenPricesCopy[symbol] *= tokenBalanceInWholeUnits);
+    return tokenPricesCopy;
+  }
+
+  addBalances(totalBalances: {[symbol: string]: number}, toAddBalances: {[symbol: string]: number}) {
+    const copyTotalBalances = clonedeep(totalBalances);
+    Object.keys(toAddBalances).map((symbol: string) => {
+      if (copyTotalBalances[symbol] === undefined || copyTotalBalances === null) {
+        copyTotalBalances[symbol] = 0;
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: throw an exception (use ensure)

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 9]
```diff
@@ -1,33 +1,82 @@
...
-  getTokenPrice: (tokenSymbol: string, currencySymbol: string) => Promise<number>;
-}
+export type SymbolToValue = {[symbol: string]: number};
 
 export class AggregateBalanceObserver {
+  private tokenPrices: {[symbol: string]: SymbolToValue} = {};
+  private tokenDetailsWithBalance: TokenDetailsWithBalance[] = [];
+  private unsubscribePriceObserver?: Function;
+  private unsubscribeBalanceObserver?: Function;
+  private callbacks: Function[] = [];
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: Optionally:

Extract `callbacks` to Listeners<T>, ie:
```ts
class Listeners<T> {
  private callbacks: (T) => void [] = [];  
  notifyListeners();
}
```

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 10]
```diff
@@ -1,33 +1,82 @@
...
     const tokenBalanceInWholeUnits = Number(utils.formatEther(tokenDetailsWithBalance.balance));
-    const price = await this.priceObserver.getTokenPrice(tokenDetailsWithBalance.symbol, currencySymbol);
-    return tokenBalanceInWholeUnits * price;
+    const tokenPrices = this.tokenPrices[tokenDetailsWithBalance.symbol];
+    const tokenPricesCopy = clonedeep(tokenPrices);
+    Object.keys(tokenPricesCopy).map((symbol) => tokenPricesCopy[symbol] *= tokenBalanceInWholeUnits);
+    return tokenPricesCopy;
+  }
+
+  addBalances(totalBalances: {[symbol: string]: number}, toAddBalances: {[symbol: string]: number}) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: That should be an util

File: universal-login-sdk/test/unit/core/AggregateBalanceObserver.ts [hunk 11]
```diff
@@ -1,48 +1,82 @@
...
+  mockedAggregateBalanceObserver.notifyListeners = () => {};
+
   const tokenDetailsWithBalance: TokenDetailsWithBalance[] = [
     {...ETHER_NATIVE_TOKEN, balance: utils.parseEther('1')},
     {address: TEST_CONTRACT_ADDRESS, symbol: 'Mock', name: 'MockToken', balance: utils.parseEther('2')}
   ];
-  const ETH_PRICE_IN_USD = 1405;
+  const tokenPrices = {
+    ETH: {USD: 218.21, EUR: 194.38, BTC: 0.01893},
+    Mock: {USD: 0.02619, EUR: 0.02337, BTC: 0.00000227}
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/unit/core/AggregateBalanceObserver.ts]:
```ts
const tokenPrices = {
    ETH: {USD: 218.21, EUR: 194.38, BTC: 0.01893},
    Mock: {USD: 0.02619, EUR: 0.02337, BTC: 0.00000227}
}

const balances: {
 ETH: 1,
 MockToken: 2
}
```

```ts
const tokenValues = {
    ETH: {USD: 218.21, EUR: 194.38, BTC: 0.01893},
    Mock: {USD: 0.04619, EUR: 0.04337, BTC: 0.00000427}
}
```

```ts
{USD: 218.21, EUR: 194.38, BTC: 0.01893}
```

File: universal-login-sdk/test/unit/core/AggregateBalanceObserver.ts [hunk 12]
```diff
@@ -1,48 +1,82 @@
...
+      const token1TotalWorth = {USD: 218.21, EUR: 194.38, BTC: 0.01893};
+      const token2TotalWorth = {USD: 1838.51, EUR: 1494.71, BTC: 0.09893};
 
-  context('token balance in USD', async () => {
-    it('ETH', async () => {
-      const ethBalanceInUSD = await aggregateBalanceObserver.getTokenBalance(tokenDetailsWithBalance[0], 'USD');
+      const tokensTotalWorth = mockedAggregateBalanceObserver.addBalances(token1TotalWorth, token2TotalWorth);
 
-      expect(ethBalanceInUSD).to.be.greaterThan(1 * ETH_PRICE_IN_USD - 0.1).and.to.be.lessThan(1 * ETH_PRICE_IN_USD + 0.1);
+      expect(tokensTotalWorth).to.deep.equal({USD: 218.21 + 1838.51, EUR: 194.38 + 1494.71, BTC: 0.01893 + 0.09893});
```
Comment (marekkirejczyk) [inline universal-login-sdk/test/unit/core/AggregateBalanceObserver.ts]: In tests, it is much easier to use simple values like 1, 2, 10, 20 and build simple examples.
Makes it more enjoyable to write and easier to read.

```ts
const token1TotalWorth = {USD: 100, EUR: 150, BTC: 0.01};
const token2TotalWorth = {USD: 1000, EUR: 1500, BTC: 0.1};
expect(tokensTotalWorth).to.deep.equal({USD: 1100, EUR: 1659, BTC: 0.11});
```

File: universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts [hunk 13]
```diff
@@ -1,33 +1,82 @@
...
-  async getTokenBalance(tokenDetailsWithBalance: TokenDetailsWithBalance, currencySymbol: string) {
+  notifyListeners() {
+    if (Object.keys(this.tokenPrices).length === 0 || this.tokenDetailsWithBalance.length === 0) {
+      return;
+    }
+    const totalWorth = this.getAggregatedTotalWorth();
+    this.callbacks.forEach((callback) => callback(totalWorth));
+  }
+
+  getAggregatedTotalWorth() {
```
Comment (marekkirejczyk) [inline universal-login-sdk/lib/core/observers/AggregateBalanceObserver.ts]: Let's refactor last 4 functions:
1) You can use mapValues from loadash
2) Introduce new types:
```ts
interface ConverstionMapping {
  symbol: [symbol: BigNumber]
}

class TokenPriceMapping {
  conversions : ConverstionMapping;
  sums(tokenBalances: []);
  total();
}

}

```

File: universal-login-commons/lib/core/services/TokensValueConverter.ts [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    return totalBalances;
+  }
+
+  getZeroedBalances() {
+    const zeroBalances = {} as CurrencyToValue;
+    for (const symbol of this.observedCurrencies) {
+      zeroBalances[symbol] = 0;
+    }
+    return zeroBalances;
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/services/TokensValueConverter.ts]: Let's use lodash's `mapValues` (or write our own):
```ts
return _.mapValues(t, () => 0)
```

File: universal-login-commons/lib/core/services/TokensValueConverter.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+
+  getTokenTotalWorth(balance: utils.BigNumber, tokenPrices: CurrencyToValue) {
+    const tokenValues = {} as CurrencyToValue;
+    for (const symbol in tokenPrices) {
+      tokenValues[symbol as ObservedCurrency] = this.safeMultiply(balance, tokenPrices[symbol as ObservedCurrency]);
+    }
+    return tokenValues;
+  }
+
+  safeMultiply(balance: utils.BigNumber, price: number) {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/services/TokensValueConverter.ts]: Extract util: `safeMultiply(left: utils.BigNumberish, right: utils.BigNumberish)`

File: universal-login-commons/lib/core/services/TokensValueConverter.ts [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+      (tokensTotal, tokenTotal) =>
+        (this.addBalances(tokensTotal, tokenTotal)), this.getZeroedBalances()
+      );
+  }
+
+  getTokensTotalWorth(tokensDetailsWithBalance: TokenDetailsWithBalance[], tokensPrices: TokensPrices) {
+    return tokensDetailsWithBalance.map((token) => this.getTokenTotalWorth(token.balance, tokensPrices[token.symbol]));
+  }
+
+  getTokenTotalWorth(balance: utils.BigNumber, tokenPrices: CurrencyToValue) {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/services/TokensValueConverter.ts]: Let's use lodash's mapValues (or write our own):
```ts
return _.mapValues(tokenPrices, (value) => safeMultiply(balance, value))
```

File: universal-login-commons/lib/core/services/TokensValueConverter.ts [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+const SAFE_MULTIPLY_UNITS_NORMALIZE = 18 + 18;
+const SAFE_MULTIPLY_UNITS_DENORMALIZE = 18;
+
+export class TokensValueConverter {
+  constructor(private observedCurrencies: ObservedCurrency[]) {}
+
+  getTotal(tokensDetailsWithBalance: TokenDetailsWithBalance[], tokensPrices: TokensPrices) {
+    const tokensTotalWorth = this.getTokensTotalWorth(tokensDetailsWithBalance, tokensPrices);
+    return tokensTotalWorth.reduce(
+      (tokensTotal, tokenTotal) =>
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/services/TokensValueConverter.ts]: (total, subTotal)

**Comment by marekkirejczyk** [comment]
Merging, put some proposed refactorings to be perhaps done next week.

**Pull request merged**